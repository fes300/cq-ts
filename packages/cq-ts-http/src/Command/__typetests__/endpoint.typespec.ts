import { Endpoint } from '..';
import * as t from 'io-ts';

Endpoint({
  Input: {
    Query: { color: t.string },
    Params: { id: t.string },
  },
  Method: 'GET',
  // @dts-jest:fail:snap accessing non defined Params in getPath is not allowed
  getPath: ({ foo }) => `users/${foo}/crayons`,
  Output: t.type({ crayons: t.array(t.string) }),
  Opts: { stringifyBody: true },
});

Endpoint({
  Input: {
    Query: { color: t.string },
    Params: { id: t.string },
  },
  Method: 'GET',
  // @dts-jest:pass:snap accessing defined Params in getPath is allowed
  getPath: ({ id }) => `users/${id}/crayons`,
  Output: t.type({ crayons: t.array(t.string) }),
  Opts: { stringifyBody: true },
});

Endpoint({
  Input: {
    Query: { color: t.string },
    Params: { id: t.string },
    // @dts-jest:fail:snap Body is not allowed with method !== POST | PUT | PATCH
    Body: t.type({ foo: t.string }),
  },
  Method: 'GET',
  getPath: ({ id }) => `users/${id}/crayons`,
  Output: t.type({ crayons: t.array(t.string) }),
  Opts: { stringifyBody: true },
});

Endpoint({
  Input: {
    Query: { color: t.string },
    Params: { id: t.string },
    // @dts-jest:pass:snap  Body is allowed with POST
    Body: t.type({ foo: t.string }),
  },
  Method: 'POST',
  getPath: ({ id }) => `users/${id}/crayons`,
  Output: t.type({ crayons: t.array(t.string) }),
  Opts: { stringifyBody: true },
});

Endpoint({
  Input: {
    Query: { color: t.string },
    Params: { id: t.string },
    // @dts-jest:pass:snap Body is allowed with PUT
    Body: t.type({ foo: t.string }),
  },
  Method: 'PUT',
  getPath: ({ id }) => `users/${id}/crayons`,
  Output: t.type({ crayons: t.array(t.string) }),
  Opts: { stringifyBody: true },
});

Endpoint({
  Input: {
    Query: { color: t.string },
    Params: { id: t.string },
    // @dts-jest:pass:snap Body is allowed with PATCH
    Body: t.type({ foo: t.string }),
  },
  Method: 'PATCH',
  getPath: ({ id }) => `users/${id}/crayons`,
  Output: t.type({ crayons: t.array(t.string) }),
  Opts: { stringifyBody: true },
});

Endpoint({
  Input: {
    Query: { color: t.string },
    Params: { id: t.string },
    // @dts-jest:pass:snap Body is allowed with unions
    Body: t.union([t.type({ foo: t.string }), t.type({ baz: t.string })]),
  },
  Method: 'PATCH',
  getPath: ({ id }) => `users/${id}/crayons`,
  // @dts-jest:pass:snap Body is allowed with unions
  Output: t.union([t.type({ foo: t.string }), t.type({ baz: t.string })]),
  Opts: { stringifyBody: true },
});

Endpoint({
  Input: {
    Query: { color: t.string },
    Params: { id: t.string },
    // @dts-jest:pass:snap Body is allowed with intersections
    Body: t.intersection([t.type({ foo: t.string }), t.type({ baz: t.string })]),
  },
  Method: 'PATCH',
  getPath: ({ id }) => `users/${id}/crayons`,
  // @dts-jest:pass:snap Body is allowed with intersections
  Output: t.union([t.type({ foo: t.string }), t.type({ baz: t.string })]),
  Opts: { stringifyBody: true },
});

const endpointInstance = Endpoint({
  Input: {
    Query: { color: t.string },
    Params: { id: t.string },
  },
  Method: 'GET',
  getPath: ({ id }) => `users/${id}/crayons`,
  Output: t.type({ crayons: t.array(t.string) }),
  Opts: { stringifyBody: true },
});

// @dts-jest:pass:snap resulting EndpointInstances typings are correct
endpointInstance.Input.Params.props.id;
// @dts-jest:pass:snap resulting EndpointInstances typings are correct
endpointInstance.Input.Query.props.color;
// @dts-jest:pass:snap resulting EndpointInstances typings are correct
endpointInstance.getStaticPath;
// @dts-jest:fail:snap resulting EndpointInstances typings are correct
endpointInstance.Input.Body?.prova;
// @dts-jest:fail:snap resulting EndpointInstances typings are correct
endpointInstance.Output.props.fakeOutput;
// @dts-jest:pass:snap resulting EndpointInstances typings are correct
endpointInstance.Output.props.crayons;

const endpointWithParam = Endpoint({
  Input: {
    Query: { color: t.string },
    Params: { id: t.number },
  },
  Method: 'GET',
  getPath: ({ id }) => `users/${id.toString()}/crayons`,
  Output: t.type({ crayons: t.array(t.string) }),
  Opts: { stringifyBody: true },
});

const endpointWithoutParam = Endpoint({
  Input: {
    Query: { color: t.string },
  },
  Method: 'GET',
  getPath: () => `users/crayons`,
  Output: t.type({ crayons: t.array(t.string) }),
  Opts: { stringifyBody: true },
});

// @dts-jest:fail:snap getPath cannot be called with no args if there are params defined in the endpoint
endpointWithParam.getPath();
// @dts-jest:fail:snap getPath cannot be called with empty object if there are Params defined in the endpoint
endpointWithParam.getPath({});
// @dts-jest:fail:snap getPath cannot be called with args different from the Params defined in the endpoint
endpointWithParam.getPath({ foo: '' });
// @dts-jest:fail:snap getPath args must have the same type as the Params defined in the endpoint
endpointWithParam.getPath({ id: '' });

// @dts-jest:pass:snap getPath can be called with no args if no Params are defined in the endpoint
endpointWithoutParam.getPath();
// @dts-jest:fail:snap getPath cannot be called with empty object if no Params are defined in the endpoint
endpointWithoutParam.getPath({});
// @dts-jest:fail:snap getPath cannot be called with any arg if no Params are defined in the endpoint
endpointWithoutParam.getPath({ id: 2 });

// @dts-jest:fail:snap getStaticPath requires a mapping function if some Params are defined in the endpoint
endpointWithParam.getStaticPath();
// @dts-jest:pass:snap getStaticPath requires a mapping function if some Params are defined in the endpoint
endpointWithParam.getStaticPath((param) => `:${param}`);

// @dts-jest:fail:snap getStaticPath requires no args if no Params are defined
endpointWithoutParam.getStaticPath((param) => `:${param}`);
// @dts-jest:pass:snap getStaticPath requires no args if no Params are defined
endpointWithoutParam.getStaticPath();
