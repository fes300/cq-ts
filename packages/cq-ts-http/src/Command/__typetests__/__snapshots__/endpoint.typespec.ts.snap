// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Body is allowed with PATCH (type) should match snapshot 1`] = `"TypeC<{ foo: StringC; }>"`;

exports[`Body is allowed with POST (type) should match snapshot 1`] = `"TypeC<{ foo: StringC; }>"`;

exports[`Body is allowed with PUT (type) should match snapshot 1`] = `"TypeC<{ foo: StringC; }>"`;

exports[`Body is allowed with intersections (type) should match snapshot 1`] = `"IntersectionC<[TypeC<{ foo: StringC; }>, TypeC<{ baz: StringC; }>]>"`;

exports[`Body is allowed with intersections (type) should match snapshot 2`] = `"UnionC<[TypeC<{ foo: StringC; }>, TypeC<{ baz: StringC; }>]>"`;

exports[`Body is allowed with unions (type) should match snapshot 1`] = `"UnionC<[TypeC<{ foo: StringC; }>, TypeC<{ baz: StringC; }>]>"`;

exports[`Body is allowed with unions (type) should match snapshot 2`] = `"UnionC<[TypeC<{ foo: StringC; }>, TypeC<{ baz: StringC; }>]>"`;

exports[`Body is not allowed with method !== POST | PUT | PATCH (type) should match snapshot 1`] = `"Type 'TypeC<{ foo: StringC; }>' is not assignable to type 'never'."`;

exports[`accessing defined Params in getPath is allowed (type) should match snapshot 1`] = `"({ id }: { id: string; }) => string"`;

exports[`accessing non defined Params in getPath is not allowed (type) should match snapshot 1`] = `"Property 'foo' does not exist on type '{ id: string; }'."`;

exports[`getPath args must have the same type as the Params defined in the endpoint (type) should match snapshot 1`] = `"Type 'string' is not assignable to type 'number'."`;

exports[`getPath can be called with no args if no Params are defined in the endpoint (type) should match snapshot 1`] = `"string"`;

exports[`getPath cannot be called with any arg if no Params are defined in the endpoint (type) should match snapshot 1`] = `"Expected 0 arguments, but got 1."`;

exports[`getPath cannot be called with args different from the Params defined in the endpoint (type) should match snapshot 1`] = `
"Argument of type '{ foo: string; }' is not assignable to parameter of type '{ id: number; }'.
  Object literal may only specify known properties, and 'foo' does not exist in type '{ id: number; }'."
`;

exports[`getPath cannot be called with empty object if no Params are defined in the endpoint (type) should match snapshot 1`] = `"Expected 0 arguments, but got 1."`;

exports[`getPath cannot be called with empty object if there are Params defined in the endpoint (type) should match snapshot 1`] = `
"Argument of type '{}' is not assignable to parameter of type '{ id: number; }'.
  Property 'id' is missing in type '{}' but required in type '{ id: number; }'."
`;

exports[`getPath cannot be called with no args if there are params defined in the endpoint (type) should match snapshot 1`] = `"Expected 1 arguments, but got 0."`;

exports[`getStaticPath requires a mapping function if some Params are defined in the endpoint (type) should match snapshot 1`] = `"Expected 1 arguments, but got 0."`;

exports[`getStaticPath requires a mapping function if some Params are defined in the endpoint (type) should match snapshot 2`] = `"string"`;

exports[`getStaticPath requires no args if no Params are defined (type) should match snapshot 1`] = `"Expected 0 arguments, but got 1."`;

exports[`getStaticPath requires no args if no Params are defined (type) should match snapshot 2`] = `"string"`;

exports[`resulting EndpointInstances typings are correct (type) should match snapshot 1`] = `"StringC"`;

exports[`resulting EndpointInstances typings are correct (type) should match snapshot 2`] = `"StringC"`;

exports[`resulting EndpointInstances typings are correct (type) should match snapshot 3`] = `"(f: (paramName: string) => string) => string"`;

exports[`resulting EndpointInstances typings are correct (type) should match snapshot 4`] = `"Property 'prova' does not exist on type 'never'."`;

exports[`resulting EndpointInstances typings are correct (type) should match snapshot 5`] = `"Property 'fakeOutput' does not exist on type '{ crayons: ArrayC<StringC>; }'."`;

exports[`resulting EndpointInstances typings are correct (type) should match snapshot 6`] = `"ArrayC<StringC>"`;
